MinhNguyen/vine_copulas.py

```diff
@@ -0,0 +1,238 @@
+import pandas as pd
+import numpy as np
+
+from itertools import combinations
+from tqdm import tqdm
+from statsmodels.distributions.empirical_distribution import ECDF
+
+from stat_utils import SpearmanExt
+
+
+class PartnerSelector():
+    """An implementation of the Partner Selection methods as described in 
+    Chapter 3.1.1 of Statistical arbitrage with vine copulas by Stubinger, J., Mangold, B., Krauss, C. (2016)
+
+    https://www.econstor.eu/bitstream/10419/147450/1/870932616.pdf
+
+    Attributes
+    ----------
+        data (pd.DataFrame): original price data input
+        data_returns (pd.DataFrame): discrete returns calculated from input data
+        data_rank (pd.DataFrame): relative ranks of `data_returns`, this is transposed
+        data_corr (pd.DataFrame): correlation matrix from relative ranks
+        dimensions (int): number of tickers to put in each combination including the target ticker
+    """
+
+    class Method():
+        """Constants for the different partner selection prodedures"""
+        TRADITIONAL = 'traditional'
+        EXTENDED = 'extended'
+        GEOMETRIC = 'geometric'
+
+
+    def __init__(self, data_prices, dimensions=4, threshold=50):
+        """Upon initialization of this object, the pre-selection algorithm will be applied to compute
+        the combinations of the top correlated symbols relative to each target ticker
+
+        :param data_prices: (pd.DataFrame): input data with tickers along the column axis
+        :param dimensions: (int): number of tickers to output in each combination
+        :param threshold: (int): the number of top correlated tickers to construct combinations from,
+                                used to reduce computational load
+        """
+        self.data = data_prices
+        data_returns = data_prices.pct_change().iloc[1:]
+        # Transpose to set index as ticker symbols for convenience
+        self.data_returns = data_returns.transpose()
+        self.data_rank = self.data_returns.rank(axis=1)
+        # Calculate correlation matrix via spearman
+        self.data_corr = data_returns.corr(method='spearman')
+        self.dimensions = dimensions
+        self._threshold = threshold
+
+        # Internal data structure, keys are target tickers and values are DataFrames
+        # containing potential combinations along with various association metrics
+        self._partner_data = {ticker:self._get_combination_df(ticker, self._threshold) for ticker in tqdm(self.data_corr.index, desc="Calculating preselections")}
+
+    def get_partners(self, method=Method.TRADITIONAL, targets=None):
```

Comment:
Hmm I saw you introduce this technique but I am not convinced that this is the best way to do it. 

Maybe you can please elaborate as to why you did it this way?

Reply:
This is quite an awkward way to call the traditional method. `selector.get_partners(method=PartnerSelector.Method.TRADITIONAL, targets=target_tickers)`.

The user needs to know your internal class structure to know that they need to call `PartnerSelector.Method.TRADITIONAL` inside this function. 

That seems like a bad idea, what am I missing?

Reply:
This seems like an artifact from C# or maybe Java.